### PR TITLE
Fixes for capture descriptors

### PR DESCRIPTION
**Pull request by slavapestov**

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@bitjammer Can you take a quick look?

File: include/swift/Remote/MetadataReader.h [hunk 1]
```diff
@@ -166,6 +181,51 @@ class TypeDecoder {
...
+            flags =
+              flags.withConvention(FunctionMetadataConvention::CFunctionPointer);
+          } else if (text == "@convention(block)") {
+            flags =
+              flags.withConvention(FunctionMetadataConvention::Block);
+          }
+        }
+      }
+
+      // Completely punt on argument types and results.
```
Comment (bitjammer) [inline include/swift/Remote/MetadataReader.h]: Should we fix this later or is this something we don't care about?

File: lib/IRGen/GenReflection.cpp [hunk 2]
```diff
@@ -550,83 +538,62 @@ class CaptureDescriptorBuilder : public ReflectionMetadataBuilder {
...
+        auto CaptureType = getElementTypes()[Index];
+        assert(ParamType == CaptureType);
+        (void) ParamType;
+        (void) CaptureType;
+
+        auto Root = SourceBuilder.createReferenceCapture(Index);
+        auto Src = Fulfillment->Path.getMetadataSource(SourceBuilder, Root);
+
+        auto SubstType = Caller.mapTypeOutOfContext(SubstMap[GenericParam]);
+        SourceMap.push_back({SubstType->getCanonicalType(), Src});
```
Comment (bitjammer) [inline lib/IRGen/GenReflection.cpp]: Nice, this is much clearer now.

File: test/Reflection/capture_descriptors.sil [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+// CHECK-NEXT:  (bound_generic_enum Swift.Optional
+// CHECK-NEXT:    (function
+// CHECK-NEXT:      (generic_type_parameter depth=0 index=1)
+// CHECK-NEXT:      (generic_type_parameter depth=0 index=2))))
+// CHECK-NEXT: - Metadata sources:
+// CHECK-NEXT: (function
+// CHECK-NEXT:   (generic_type_parameter depth=0 index=1)
+// CHECK-NEXT:   (generic_type_parameter depth=0 index=2))
+// CHECK-NEXT: (generic_argument index=0
+// CHECK-NEXT:   (reference_capture index=0))
```
Comment (bitjammer) [inline test/Reflection/capture_descriptors.sil]: Is this reference capture the `@thick Optional<T>.Type`? Seems like a weird name for it, if so. Is this basically "pointer capture argument" now, and you have to check the typeref of the capture to know whether it's a class instance or a type?

**Comment by bitjammer** [comment]
Looks great.

**Comment by tkremenek** [comment]
@swift-ci test

File: include/swift/Remote/MetadataReader.h (hunk 1 above)
Comment (slavapestov) [inline include/swift/Remote/MetadataReader.h]: I don't think we care about it unless at some point we add support for function reflection and dynamic calls...

**Comment by slavapestov** [comment]
@swift-ci Please test

File: test/Reflection/capture_descriptors.sil (hunk 3 above)
Comment (slavapestov) [inline test/Reflection/capture_descriptors.sil]: Let's discuss this in person.

**Pull request closed without merging**